module/zfs/btree.c

```diff
@@ -0,0 +1,2191 @@
+
+		old_node->bth_parent = new_node->bth_parent = new_root;
+		new_root->btc_children[0] = old_node;
+		new_root->btc_children[1] = new_node;
+		bcopy(buf, new_root->btc_elems, size);
+
+		tree->bt_height++;
+		tree->bt_root = new_root_hdr;
+#ifdef ZFS_DEBUG
+		btree_poison_node(tree, new_root_hdr);
+#endif
+		return;
+	}
+
+	/*
+	 * Since we have the new separator, binary search for where to put
+	 * new_node.
+	 */
+	btree_index_t idx;
+	ASSERT(par_hdr->bth_core);
+	VERIFY3P(btree_find_in_buf(tree, parent->btc_elems, par_hdr->bth_count,
+	    buf, &idx), ==, NULL);
+	ASSERT(idx.bti_before);
+	uint64_t offset = idx.bti_offset;
+	ASSERT3U(offset, <=, par_hdr->bth_count);
+	ASSERT3P(parent->btc_children[offset], ==, old_node);
+
+	/*
+	 * If the parent isn't full, shift things to accomodate our insertions
+	 * and return.
+	 */
+	if (par_hdr->bth_count != BTREE_CORE_ELEMS) {
+		if (btree_verify_intensity >= 5) {
+			btree_verify_poison_at(tree, par_hdr,
+			    par_hdr->bth_count);
+		}
+
+		/* Shift existing elements and children */
+		uint64_t count = par_hdr->bth_count - offset;
+		bt_shift_core_right(tree, parent, offset, count,
+		    BSS_PARALLELOGRAM);
+
+		/* Insert new values */
+		parent->btc_children[offset + 1] = new_node;
+		bcopy(buf, parent->btc_elems + offset * size, size);
+		par_hdr->bth_count++;
+		return;
+	}
+
+	/*
+	 * We need to split this core node into two. Currently there are
+	 * BTREE_CORE_ELEMS + 1 child nodes, and we are adding one for
+	 * BTREE_CORE_ELEMS + 2. Some of the children will be part of the
+	 * current node, and the others will be moved to the new core node.
+	 * There are BTREE_CORE_ELEMS + 1 elements including the new one. One
+	 * will be used as the new separator in our parent, and the others
+	 * will be split among the two core nodes.
+	 *
+	 * Usually we will split the node in half evenly, with
+	 * BTREE_CORE_ELEMS/2 elements in each node. If we're bulk loading, we
+	 * instead move only about a quarter of the elements (and children) to
+	 * the new node. Since the average state after a long time is a 3/4
+	 * full node, shortcutting directly to that state improves efficiency.
+	 */
+	uint64_t move_count = MAX(BTREE_CORE_ELEMS / (tree->bt_bulk == NULL ?
+	    2 : 4), 2);
+	uint64_t keep_count = BTREE_CORE_ELEMS - move_count;
+	ASSERT3U(BTREE_CORE_ELEMS - move_count, >=, 2);
+	tree->bt_num_nodes++;
+	btree_core_t *new_parent = kmem_alloc(sizeof (btree_core_t) +
+	    BTREE_CORE_ELEMS * size, KM_SLEEP);
+	btree_hdr_t *new_par_hdr = &new_parent->btc_hdr;
+	new_par_hdr->bth_parent = par_hdr->bth_parent;
+	new_par_hdr->bth_core = B_TRUE;
+	new_par_hdr->bth_count = move_count;
+#ifdef ZFS_DEBUG
+	btree_poison_node(tree, new_par_hdr);
+#endif
+	par_hdr->bth_count = keep_count;
+
+	/*
+	 * The three cases to consider are that the element in buf should be
+	 * in the existing node (with lower values), the new node (with higher
+	 * values), or that it should separate the two nodes.
```

Comment:
Could we simplify this code by treating it as two steps:
1. split the parent into two nodes
2. add the element to the appropriate node (could be like `goto top`, or move the "parent isn't full" code to its own function)
This would mean an additional bcopy in some cases, but I think this is a rare case anyway - it should only happen on average every BTREE_CORE_ELEMS times, right?

Reply:
We could. I'm hesitant to do so, partly because of the additional bcopy, and partly because of precedent. Right now, all the copies are basically optimal. Nothing is moved twice except for single elements being moved a temporary buffer before being moved to separators. If we're going to accept redundant moves, there's a number of places in the code that we could do so where it would simplify the logic somewhat (both insert and both remove functions have a flow like this)

Reply:
Given that you've already written the code, it seems like the considerations are the future maintainability of the code, the performance of the code, and the amount of work we need to do now.  In terms of precedent, I think we'd want to make similar changes everywhere that we can simplify the number of cases.  Are there particular workloads where the additional bcopy's would have a significant overhead?

Reply:
If we were to apply it in each situation, we would have redundant bcopys about half of the time when we split or merge any node (leaf or core). These operations are relatively rare in normal operation (once every BTREE_LEAF_SIZE/elem size operations, roughly). It's possible to come up with simple scenarios that would make it happen more often, but none seem likely to occur in the range tree model (some might occur in the size-sorted trees, but still probably not very often).

Making the change would involve substantial additional testing, but would hopefully ease the readability and maintainability of the code going forwards. If we're willing to accept the performance costs, it's probably for the best.

Reply:
I see, the requirement of having all nodes (except for root) be at least half full means that we have to be careful about where we split this node so that the new entry goes into the "smaller" new node, bringing it up to exactly half full.

Why do we have the requirement of being half full, again?  Does that requirement mean that you could get into a situation where you are alternatively adding and removing one entry, and every operation causes us to either split or join nodes?  Could we relax the requirement so that it only needs to be 1/3 full, so that we could tolerate some removals after a split, without having to rejoin?  (Though I imagine this could get complicated, it might be mitigated by simplifying the splitting code.)

Reply:
Allowing fills less than 50% results in situations where we can significant inflate memory usage. Since our memory usage is based on the average fill rate, dropping the floor can cause significantly reduced efficiency.

The situation you describe is possible but unlikely; the node we're removing from and both of its neighbors have to all be at exactly 50% capacity. We could also alleviate this by adding support for "giving" an extra element to a neighbor when a node is about to become too large, in addition to "taking" from them when a node is about to become underfull. If that functionality was added, the only way to get the situation you describe is if the tree has exactly enough elements to fill a leaf node, plus one (so we'd be constantly splitting and merging the only two children of the root).